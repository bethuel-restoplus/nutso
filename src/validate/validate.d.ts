import { Schema } from "../schema/Schema";
import { Result } from "../result/Result";
import { FieldPath } from "../models/FieldPath";
export declare const validate: <T>(o: T, schema: Schema<T>, fieldPath?: FieldPath) => Result<T>;
