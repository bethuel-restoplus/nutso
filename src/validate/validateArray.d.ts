import { ArrayResult } from "../result/ArrayResult";
import { ArraySchema } from "../schema/ArraySchema";
import { FieldPath } from "../models/FieldPath";
export declare const validateArray: <T>(arr: T[], schema: ArraySchema<T>, fieldPath: FieldPath) => ArrayResult<T>;
