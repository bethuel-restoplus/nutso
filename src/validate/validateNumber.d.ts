import { NumberSchema } from "../schema/NumberSchema";
import { FieldPath } from "../models/FieldPath";
export declare const validateNumber: (o: any, schema: NumberSchema, fieldPath: FieldPath) => import("../result/BaseResult").BaseResult;
