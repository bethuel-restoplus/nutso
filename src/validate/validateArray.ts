import { ArrayResult } from "../models/result/ArrayResult";
import { ArraySchema } from "../models/schema/ArraySchema";
import { FieldPath } from "../models/FieldPath";
import { isNil } from "../utils/typeChecker";
import { _validate } from "./validate";
import { optionalFlagValidator } from "../utils/optionalFlagValidator";

export const validateArray = <T, R>(arr: T[], root: R, schema: ArraySchema<T, R>): ArrayResult<T> => {
  // isnil
  if (isNil(arr)) {
    return { ...optionalFlagValidator(root, schema.optional), items: [] };
  }

  const result: ArrayResult<T> = {
    isValid: true,
    errorMessage: "",
    items: [],
    errorPath: [],
  };

  // array min-items
  if (!isNil(schema.minItems) && arr.length < schema.minItems!) {
    result.isValid = false;
    result.errorMessage = `Should have at least ${schema.minItems} items.`;
  }

  // array max-items
  if (!isNil(schema.maxItems) && arr.length >= schema.maxItems!) {
    result.isValid = false;
    result.errorMessage = `Should not have more than ${schema.maxItems! - 1} items.`;
  }

  // for each key, validate
  for (let i = 0; i < arr.length; i++) {
    result.items[i] = _validate(arr[i], root, schema.items);
  }

  // if this node is valid, then check if all of it's children are valid
  // because the node is invalid, if any of it's children are invalid
  if (result.isValid) {
    for (let i = 0; i < arr.length; i++) {
      const item = result.items[i];
      if (!item.isValid) {
        result.isValid = false;
        result.errorMessage = item.errorMessage;
        result.errorPath = [i, ...item.errorPath];
        break;
      }
    }
  }

  return result;
};
