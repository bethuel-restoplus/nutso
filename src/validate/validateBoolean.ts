import { BooleanResult } from "../models/result/BooleanResult";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { FieldPath } from "../models/FieldPath";
import { isBoolean, isNil } from "../utils/typeChecker";

export const validateBoolean = (o: any, schema: BooleanSchema): BooleanResult => {
  //

  // isnil
  if (isNil(o)) {
    if (!schema.optional) {
      return {
        isValid: false,
        errorMessage: `Required field.`,
        errorPath: []
      };
    }
    return {
      isValid: true,
      errorMessage: "",
      errorPath: []
    };
  }

  if (!isBoolean(o)) {
    return {
      isValid: false,
      errorMessage: `Should be true or false.`,
      errorPath: []
    };
  }

  return {
    isValid: true,
    errorMessage: ``,
    errorPath: []
  };
};
