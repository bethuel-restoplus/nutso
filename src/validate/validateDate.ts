import { DateResult } from "../models/result/DateResult";
import { DateSchema } from "../models/schema/DateSchema";
import { isDate, isNil } from "../utils/typeChecker";

export const validateDate = (o: any, schema: DateSchema): DateResult => {
  //

  // isnil
  if (isNil(o)) {
    if (!schema.optional) {
      return {
        isValid: false,
        errorMessage: `Required field.`
      };
    }
    return {
      isValid: true,
      errorMessage: ""
    };
  }

  if (!isDate(o)) {
    return {
      isValid: false,
      errorMessage: `Should be a valid date.`
    };
  }

  return {
    isValid: true,
    errorMessage: ``
  };
};
