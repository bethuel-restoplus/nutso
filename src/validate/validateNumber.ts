import { NumberResult } from "../models/result/NumberResult";
import { NumberSchema } from "../models/schema/NumberSchema";
import { FieldPath } from "../models/FieldPath";
import { isNil, isNumber } from "../utils/typeChecker";

export const validateNumber = (o: any, schema: NumberSchema): NumberResult => {
  //

  // isnil
  if (isNil(o)) {
    if (!schema.optional) {
      return {
        isValid: false,
        errorMessage: `Required field.`
      };
    }
    return {
      isValid: true,
      errorMessage: ""
    };
  }

  if (!isNumber(o)) {
    return {
      isValid: false,
      errorMessage: `Should be a number.`
    };
  }

  const numbr = o as number;

  // min length
  if (!isNil(schema.min) && numbr < schema.min!) {
    return {
      isValid: false,
      errorMessage: `Should not be less than ${schema.min}.`
    };
  }

  // max length
  if (!isNil(schema.max) && numbr > schema.max!) {
    return {
      isValid: false,
      errorMessage: `Should not be larger than ${schema.max}.`
    };
  }

  return {
    isValid: true,
    errorMessage: ``
  };
};
