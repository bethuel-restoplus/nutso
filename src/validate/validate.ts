import { Schema } from "../models/schema/Schema";
import { Result } from "../models/result/Result";
import { validateString } from "./validateString";
import { StringSchema } from "../models/schema/StringSchema";
import { NumberSchema } from "../models/schema/NumberSchema";
import { validateNumber } from "./validateNumber";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { validateBoolean } from "./validateBoolean";
import { ObjectSchema } from "../models/schema/ObjectSchema";
import { validateObject } from "./validateObject";
import { FieldPath } from "../models/FieldPath";
import { validateArray } from "./validateArray";
import { ArraySchema } from "../models/schema/ArraySchema";
import { validateDate } from "./validateDate";
import { DateSchema } from "../models/schema/DateSchema";

export const validate = <T, R>(o: T | null, root: R, schema: Schema<T, R>): Result<T> => {
  //
  if (!schema) {
    throw new Error("Schema should not be null");
  }

  /**
   * validate object against the schema
   */
  switch (schema.type) {
    case "number":
      return validateNumber(o, schema as NumberSchema) as Result<T>;
    case "string":
      return validateString(o, root, schema as StringSchema<R>) as Result<T>;
    case "boolean":
      return validateBoolean(o, schema as BooleanSchema) as Result<T>;
    case "array":
      return validateArray(o as any, root, schema as ArraySchema<T, R>) as Result<T>;
    case "object":
      return validateObject(o, root, schema as ObjectSchema<T, R>) as Result<T>;
    case "date":
      return validateDate(o, schema as DateSchema) as Result<T>;
  }
  throw new Error(`Unhandled data type`);
};
