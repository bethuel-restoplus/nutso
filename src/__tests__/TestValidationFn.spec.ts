import { Schema } from "..";
import { validate } from "../validate/validate";

test(`Compose schema with different roots - typesafety`, () => {
  type Address = {
    line1: string;
  };
  type Customer = {
    name: string;
    address: Address;
  };

  const addressSchema: Schema<Address, Customer> = {
    type: "object",
    properties: {
      line1: {
        type: "string",
        validatorFn: (value, root) => {
          root.address.line1;
          return {
            isValid: true,
            errorMessage: ""
          };
        }
      }
    }
  };

  const customerSchema: Schema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        validatorFn: (value, root) => {
          return {
            isValid: false,
            errorMessage: "Custom validation failed"
          };
        }
      },
      address: addressSchema
    }
  };

  // this test is just to ensure typescript is not complaining!
  // so no assertion required
  expect(true).toBeTruthy();
});

test(`Custom validation - check errorPath - level 1`, () => {
  type Address = {
    line1: string;
  };
  type Customer = {
    name: string;
    address: Address;
  };

  const addressSchema: Schema<Address, Customer> = {
    type: "object",
    properties: {
      line1: {
        type: "string",
        validatorFn: (value, root) => {
          root.address.line1;
          return {
            isValid: true,
            errorMessage: ""
          };
        }
      }
    }
  };

  const customerSchema: Schema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        validatorFn: (value, root) => {
          return {
            isValid: false,
            errorMessage: "Custom validation failed"
          };
        }
      },
      address: addressSchema
    }
  };

  const customer: Customer = {
    name: "John",
    address: {
      line1: "10 Downing St"
    }
  };

  const result = validate(customer, customer, customerSchema);
  expect(result).toMatchSnapshot();
});

test(`Custom validation - check errorPath - level 2`, () => {
  type Address = {
    line1: string;
  };
  type Customer = {
    name: string;
    address: Address;
  };

  const addressSchema: Schema<Address, Customer> = {
    type: "object",
    properties: {
      line1: {
        type: "string",
        validatorFn: (value, root) => {
          root.address.line1;
          return {
            isValid: false,
            errorMessage: "Custom validation failed"
          };
        }
      }
    }
  };

  const customerSchema: Schema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        validatorFn: (value, root) => {
          return {
            isValid: true,
            errorMessage: ""
          };
        }
      },
      address: addressSchema
    }
  };

  const customer: Customer = {
    name: "John",
    address: {
      line1: "10 Downing St"
    }
  };

  const result = validate(customer, customer, customerSchema);
  expect(result).toMatchSnapshot();
});
