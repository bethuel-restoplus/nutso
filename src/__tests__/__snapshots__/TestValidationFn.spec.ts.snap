// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom validation - check errorPath - level 1 1`] = `
Object {
  "errorMessage": "Custom validation failed",
  "errorPath": Array [
    "name",
  ],
  "isValid": false,
  "properties": Object {
    "address": Object {
      "errorMessage": "",
      "errorPath": Array [],
      "isValid": true,
      "properties": Object {
        "line1": Object {
          "errorMessage": "",
          "errorPath": Array [],
          "isValid": true,
        },
      },
    },
    "name": Object {
      "errorMessage": "Custom validation failed",
      "errorPath": Array [],
      "isValid": false,
    },
  },
}
`;

exports[`Custom validation - check errorPath - level 2 1`] = `
Object {
  "errorMessage": "Custom validation failed",
  "errorPath": Array [
    "address",
    "line1",
  ],
  "isValid": false,
  "properties": Object {
    "address": Object {
      "errorMessage": "Custom validation failed",
      "errorPath": Array [
        "line1",
      ],
      "isValid": false,
      "properties": Object {
        "line1": Object {
          "errorMessage": "Custom validation failed",
          "errorPath": Array [],
          "isValid": false,
        },
      },
    },
    "name": Object {
      "errorMessage": "",
      "errorPath": Array [],
      "isValid": true,
    },
  },
}
`;
